gnucash/gnome-utils/gnc-main-window.cpp

```diff
@@ -142,6 +146,8 @@ static QofLogModule log_module = GNC_MOD_GUI;
 static GObjectClass *parent_class = nullptr;
 /** An identifier that indicates a "main" window. */
 static GQuark window_type = 0;
+/** A hash table of the display items to update the menu and tool bars */
+static GHashTable *display_item_hash; //FIXMEb added
```

Comment:
Similar to the merged_actions_table in gnc-plugin I doubt this extra hash is really needed. From what I gather this is all about widget related properties (labels, icon names,...). So instead of storing them in a hash, why not set these properties directly on the widgets as they are created ? If you can't set them to the exact widget property immediately, you can still add them with g_object_set_data_full as generic associated data to query back later. This will save us quite a bit of duplicate hash table queries imo.

The trick will be to figure our where exactly this is best done. I suggest to find where the widgets are created (probably when loading a ui file) and and that point set the display properties related to those widgets.

Reply:
I am not sure about this, see comment below. 
In a sense I am, when the menu parts are loaded from the builder object and the menu created, the menu is scanned and the menu items updated.

Reply:
Ok, then why the intermediary step via a hash ? The properties to set on the menu items are already in a static array.

Also are the same menu items created more than once per main window ? And how about setting the properties for these menu items, are these updates frequently or only once per main window ? For example are the items recreated or the properties set when changing plugin pages ? A hash table would be useful if setting these properties had to be done over and over again.

Reply:
I think the main reason for this intermediary step was that I am using `gnc_main_window_merge_actions` which is called when the plugin is added to the window and this is all part of creating the window. So when the plugin is added, the menu item properties are added to the hash table, I think there were around 150 and then subsequently used to update the menu. Doing it this way seemed like a good idea as I would not have to fetch the properties from different plugins when required.

I do not think there were any menu items that appeared in more than one location apart from being on the toolbar.

The `File`, `Business`, `Tools`, `Reports` and `Help` are mainly static with some items being shown/hidden when required.
The `Transaction` sub menu is static but the labels are changed when on transactions or splits and only shown for registers.
The `Schedule` sub menu is static but only shown for the schedule plugin.
The `Edit`, `View` and `Action` are the ones that change the most depending on plugin.

So the main part of the menu bar is created once and updated. The `Edit`, `View`, and `Action` are created and updated on plugin focus. It may be possible to keep these sub menus once created as part of the plugin but have not looked at that which may also apply to the right mouse popup menu that is also defined in the ui files, created and updated when required.